pkg/compact/dedup/reader.go

```diff
@@ -0,0 +1,405 @@
+		samples = append(samples, ss...)
+	}
+	if len(samples) == 0 {
+		return nil, nil
+	}
+	result := make(map[SampleType][]*Sample)
+	result[RawSample] = samples
+	return result, nil
+}
+
+func (r *SampleReader) readDownSamples(tw *TimeWindow) (map[SampleType][]*Sample, error) {
+	result := make(map[SampleType][]*Sample)
+	result[CountSample] = make([]*Sample, 0)
+	result[SumSample] = make([]*Sample, 0)
+	result[MinSample] = make([]*Sample, 0)
+	result[MaxSample] = make([]*Sample, 0)
+	result[CounterSample] = make([]*Sample, 0)
+	for _, c := range r.chks {
+		chk, err := r.cr.Chunk(c.Ref)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get downsample chunk %d for labels %s", c.Ref, r.lset)
+		}
+		for _, at := range downsampleAggrTypes {
+			ac, err := chk.(*downsample.AggrChunk).Get(at)
+			if err == downsample.ErrAggrNotExist {
+				continue
+			}
+			if ac == nil {
+				level.Warn(r.logger).Log("msg", "find empty downsample chunk", "type", at, "ref", c.Ref, "labels", r.lset)
+				continue
+			}
+			samples := r.parseSamples(ac, tw)
+			if len(samples) == 0 {
+				continue
+			}
+			result[typeMapping[at]] = append(result[typeMapping[at]], samples...)
+		}
+	}
+	return result, nil
+}
+
+func (r *SampleReader) parseSamples(c chunkenc.Chunk, tw *TimeWindow) []*Sample {
+	samples := make([]*Sample, 0)
+	iterator := c.Iterator()
+	for iterator.Next() {
+		timestamp, value := iterator.At()
+		if timestamp < tw.MinTime {
+			continue
+		}
+		// Ignore the data point which timestamp is same with MaxTime.
+		// Make sure the block use scope [MinTime, MaxTime) instead of [MinTime, MaxTime]
+		if timestamp >= tw.MaxTime {
+			break
+		}
+		samples = append(samples, &Sample{
+			timestamp: timestamp,
+			value:     value,
+		})
+	}
+	return samples
+}
+
+type SampleType uint8
+
+const (
+	// samples from non-downsample blocks
+	RawSample SampleType = iota
+	// samples from downsample blocks, map to downsample AggrType
+	CountSample
+	SumSample
+	MinSample
+	MaxSample
+	CounterSample
+)
+
+var (
+	typeMapping = map[downsample.AggrType]SampleType{
+		downsample.AggrCount:   CountSample,
+		downsample.AggrSum:     SumSample,
+		downsample.AggrMin:     MinSample,
+		downsample.AggrMax:     MaxSample,
+		downsample.AggrCounter: CounterSample,
+	}
+	downsampleAggrTypes = []downsample.AggrType{
+		downsample.AggrCount,
+		downsample.AggrSum,
+		downsample.AggrMin,
+		downsample.AggrMax,
+		downsample.AggrCounter,
+	}
+)
+
+func (t SampleType) String() string {
+	switch t {
+	case RawSample:
+		return "raw"
+	case CountSample:
+		return "count"
+	case SumSample:
+		return "sum"
+	case MinSample:
+		return "min"
+	case MaxSample:
+		return "max"
+	case CounterSample:
+		return "counter"
+	}
+	return "<unknown>"
+}
+
+type BlockReader struct {
+	logger  log.Logger
+	closers []io.Closer
+
+	ir tsdb.IndexReader
+	cr tsdb.ChunkReader
+
+	postings index.Postings
+}
+
+func NewBlockReader(logger log.Logger, resolution int64, blockDir string) (*BlockReader, error) {
```

Comment:
Why we need this? We already have `tsdb.BlockReader`

Reply:
When try to get the sorted postings for one block, it needs to call four functions to make it: `OpenBlock(), Index(), Postings(), SortedPostings()`. Here just use this `BlockReader ` to provide a convenient way to access one block's postings and chunk reader.